src/app/needs/next-actions/components/NextActionsDisplay.tsx

```diff
@@ -0,0 +1,212 @@
+  const [chainEnd, setChainEnd] = useState(0); 
+  // chainEnd increments after toggling actions to refetch and update UI
+
+  useEffect(() => {
+    async function fetchData() {
+      // 1. Fetch all needs (as RxDocuments), then to JSON
+      const needsDocs = await database.getFromDb("needs");
+      const allNeeds = needsDocs.map(doc => doc.toJSON() as NeedDocument);
+
+      // 2. Highlighted needs check: selectedExpiry in the future means highlighted
+      const now = new Date();
+      const filteredNeeds = allNeeds.filter(need => {
+        const expiry = new Date(need.selectedExpiry);
+        return expiry > now && need.priority && need.priority.order > 0;
+      });
+
+      setHighlightedNeeds(filteredNeeds);
+
+      // 3. Fetch all next actions (as RxDocuments), then to JSON
+      const nextActionsDocs = await database.getFromDb("next_actions");
+      const allNextActions = nextActionsDocs.map(doc => doc.toJSON() as NextActionDocument);
+
+      // 4. Filter next actions to only those related to highlighted needs
+      const highlightedNeedIds = new Set(filteredNeeds.map(need => need.id));
+      const filteredNextActions = allNextActions.filter(na => highlightedNeedIds.has(na.need));
+
+      setRelatedNextActions(filteredNextActions);
+    }
+
+    fetchData();
+  }, [database, chainEnd]);
+
+  // Group and sort highlighted needs by priority
+  const priorityGroups = useMemo(() => {
+    if (highlightedNeeds.length === 0) return [];
+
+    const needsWithPriority = highlightedNeeds.filter(
+      need => need.priority && need.priority.order
+    );
+
+    const sortedNeeds = needsWithPriority.sort((a, b) =>
+      a.priority!.order - b.priority!.order
+    );
+
+    const groupsMap: Record<number, { priority: { order: number; name: string}; needs: NeedDocument[] }> = {};
+    for (const need of sortedNeeds) {
+      const order = need.priority!.order;
+      if (!groupsMap[order]) {
+        groupsMap[order] = {
+          priority: need.priority!,
+          needs: []
+        };
+      }
+      groupsMap[order].needs.push(need);
+    }
+
+    return Object.values(groupsMap);
+  }, [highlightedNeeds]);
+
+  // Helper to get next actions for a given need
+  const getActionsForNeed = (needId: string) => {
+    return relatedNextActions.filter(action => action.need === needId);
+  };
+
+  // onToggleAction logic
+  const onToggleAction = async (action: NextActionDocument) => {
+    const highlighted = new Date(action.selectedExpiry) > new Date(action.timestamp);
+    const collectionName = "next_actions";
+
+    if (highlighted) {
+      // Un-highlight action:
+      // Remove last selectedTimestamp
+      const updatedTimestamps = [...action.selectedTimestamps];
+      updatedTimestamps.pop();
+      
+      // Set selectedExpiry back to the original timestamp
+      await database.updateDocument(
+        collectionName,
+        action.id,
+        "selectedTimestamps",
+        updatedTimestamps
+      );
+
```

Comment:
```suggestion
```
As above, we do not need to update the selectedTimestamps when a user deselects, only the selectedExpiry

Reply:
Disagree, I think a deselection should remove the timestamp created by selecting it. Otherwise future insights analysis will show a lot of false data